app/uk/gov/hmrc/nationalimportdutyadjustmentcentrefrontend/controllers/makeclaim/UploadFormSummaryController.scala

```diff
@@ -52,7 +53,19 @@ class UploadFormSummaryController @Inject() (
     data.getAnswers map { answers =>
       Redirect(navigator.nextPage(UploadSummaryPage, answers))
     }
+  }
 
+  def onRemove(documentReference: String): Action[AnyContent] = identify.async { implicit request =>
+    data.updateAnswers(removeDocument(documentReference)) map { _ =>
+      Redirect(controllers.makeclaim.routes.UploadFormSummaryController.onPageLoad())
+    }
   }
 
+  def removeDocument: String => UserAnswers => UserAnswers = (ref: String) =>
+    (userAnswers: UserAnswers) => {
+      val uploadedFiles  = userAnswers.uploads.getOrElse(Seq.empty)
+      val remainingFiles = uploadedFiles.filterNot(file => file.upscanReference == ref)
```

Comment:
OBSERVATION: Not sure if this is any cleaner, but you could do away with the intermediate `uploadedFiles: Option[_]` via

```suggestion
      val remainingFiles = userAnswers.uploads.map(_.filterNot(_.upscanReference == ref))
```

and `remainingFiles` will already be an optional...

Reply:
Implemented your suggestion